src/assets/TokenListController.ts

```diff
@@ -0,0 +1,147 @@
+
+const DEFAULT_INTERVAL = 180 * 1000;
+
+type Token = {
+  address: string;
+  decimals: number;
+  symbol: string;
+  occurances: number;
+  aggregators: string[];
+};
+export type TokenMap = {
+  [address: string]: Token;
+};
+export type TokenListState = {
+  tokens: TokenMap;
+};
+const name = 'TokenListController';
+
+export type TokenListStateChange = {
+  type: `${typeof name}:stateChange`;
+  payload: [TokenListState, Patch[]];
+};
+
+export type GetTokenListState = {
+  type: `${typeof name}:getState`;
+  handler: () => TokenListState;
+};
+const metadata = {
+  tokens: { persist: true, anonymous: true },
+};
+const defaultState: TokenListState = {
+  tokens: {},
+};
+/**
+ * Controller that passively polls on a set interval for the list of tokens from metaswaps api
+ */
+export class TokenListController extends BaseController<
+  typeof name,
+  TokenListState
+> {
+  private mutex = new Mutex();
+
+  private intervalId?: NodeJS.Timeout;
+
+  private intervalDelay: number;
+
+  /**
+   * Creates a TokenListController instance
+   *
+   * @param options - Constructor options
+   * @param options.interval - The polling interval, in milliseconds
+   * @param options.messenger - A reference to the messaging system
+   * @param options.state - Initial state to set on this controller
+   */
+  constructor({
+    interval = DEFAULT_INTERVAL,
+    messenger,
+    state,
+  }: {
+    interval?: number;
+    messenger: RestrictedControllerMessenger<
+      typeof name,
+      GetTokenListState,
+      TokenListStateChange,
+      never,
+      never
+    >;
+    state?: Partial<TokenListState>;
+  }) {
+    super({
+      name,
+      metadata,
+      messenger,
+      state: { ...defaultState, ...state },
+    });
+    this.intervalDelay = interval;
+  }
+
+  /**
+   * Start polling for the token list
+   */
+  async start() {
+    await this.startPolling();
+  }
+
+  /**
+   * Stop polling for the token list
+   */
+  stop() {
+    this.stopPolling();
+  }
+
+  /**
+   * Prepare to discard this controller.
+   *
+   * This stops any active polling.
+   */
+  destroy() {
+    super.destroy();
+    this.stopPolling();
+  }
+
+  private stopPolling() {
+    if (this.intervalId) {
+      clearInterval(this.intervalId);
+    }
+  }
+
+  /**
+   * Starts a new polling interval
+   */
+  private async startPolling(): Promise<void> {
+    await safelyExecute(() => this.fetchTokenList());
+    this.intervalId = setInterval(async () => {
+      await safelyExecute(() => this.fetchTokenList());
+    }, this.intervalDelay);
+  }
+
+  async fetchTokenList(): Promise<void> {
+    const releaseLock = await this.mutex.acquire();
+    const { tokens }: { tokens: TokenMap } = this.state;
+    try {
+      const tokenList: Token[] = await safelyExecute(() => fetchTokenList());
+
+      for (const token of tokenList) {
+        if (!tokens[token.address]) {
+          tokens[token.address] = token;
+        }
+      }
```

Comment:
Since the source of truth is now an API, metadata could switch more often than a static json I think we need to consider a way to update the object as well.
We are persisting fields like `Token.occurances` and `Token.aggregators`, which right now do not have impact other than swaps, but do incur into changes on the UI such as alerts/warnings. I foresee we might rely on this data in other parts of the products.

Has this been deliberately ignored for this initial effort?